systemd_ctypes/bus.py

```diff
@@ -331,162 +332,491 @@ def add_object(self, path, obj):
-                    reply = message.new_method_return()
-                    reply.open_container(ord('a'), '{sv}')
-                    for name, info in self._dbus_members['properties'].items():
-                        reply.open_container(ord('e'), 'sv')
-                        reply.append_arg('s', name)
-                        self.append_property(reply, info)
-                        reply.close_container()
-                    reply.close_container()
-                    return reply.send()
+                return self.org_freedesktop_DBus_Properties_GetAll(message, *message.get_body())
 
             elif method == 'Get' and message.has_signature('ss'):
-                iface, name = message.get_body()
-                if iface == self._dbus_interface:
-                    info = self._dbus_members['properties'].get(name)
-                    if info:
-                        reply = message.new_method_return()
-                        self.append_property(reply, info)
-                        return reply.send()
-
-        elif interface == self._dbus_interface:
-            info = self._dbus_members['methods'].get(message.get_member())
-            if info:
-                handler, (out_types, in_types) = info
-                if message.has_signature(''.join(in_types)):
-                    try:
-                        result = handler(self, *message.get_body())
-                    except BusError as exc:
-                        return message.reply_method_error(exc)
-
-                    return message.reply_method_function_return_value(out_types, result)
-
-        return False
+                return self.org_freedesktop_DBus_Properties_Get(message, *message.get_body())
+
+            elif method == 'Set' and message.has_signature('ssv'):
+                return self.org_freedesktop_DBus_Properties_Get(message, *message.get_body())
+
+        try:
+            method = self._dbus_members[interface]['methods'][method]
+        except KeyError:
+            return False
+        return method.invoke(self, message)
+
+    @classmethod
+    def org_freedesktop_DBus_Introspectable_Introspect(cls, message):
+        message.reply_method_return('s', introspection.to_xml(cls._dbus_members or {}))
+
+    def org_freedesktop_DBus_Properties_PropertiesChanged(self, interface_name: str,
+                                                          changed_properties: list[dict[str, dict[str, Any]]],
+                                                          invalidated_properties: list[str]) -> None:
+        self.emit_signal('PropertiesChanged', 'sa{sv}as',
+                         interface_name, changed_properties, invalidated_properties,
+                         interface='org.freedesktop.DBus.Properties')
+
+    def org_freedesktop_DBus_Properties_Get(self, message, interface, name):
+        try:
+            prop = self._dbus_members[interface]['properties'][name]
+        except KeyError:
+            return False
```

Comment:
Shouldn't this be some concrete error reply? (Instead of trying to find a different handler and then failing with UnknownMethod)

Reply:
This got reworked.